head_first_design_patterns/tbruinem/json/src/parser.cpp

```diff
@@ -0,0 +1,196 @@
+#include "parser.hpp"
+#include "parsed_json.hpp"
+#include <stdexcept>
+#include <assert.h>
+
+namespace json
+{
+
+static bool IsWhitespace(char c) {
+	return c == ' ' || c == '\t' || c == '\n';
+}
+
+void Parser::UpdateState(ParseState state)
+{
+	this->previous_state = current_state;
+	current_state = state;
+}
+
+StateResult	Parser::StateOut(char c) {
+	//! Still out
+	if (IsWhitespace(c)) {
+		return StateResult::CONTINUE;
+	}
+	//! Within an array, we only have values
+	if (parent == ElementType::JSON_ARRAY) {
+		product = new ParsedJSON();
+		UpdateState(ParseState::VALUE);
+		if (c == '"') {
+			product->SetType(ElementType::JSON_STRING);
```

Comment:
Is this necessary? Maybe for future iterations but it seems to me that calling `StateValue(c)` will have the same effect?

Reply:
I think you're right